Functional test rework

keszybz opened the pull request:
For #13802.

keszybz commented:
Each test individually is easy, but we have so many of them that any kind of comprehensive rework takes a lot of time. The main goal of this patchset is to make things uniform, so that future changes can be done easier.

I have one more change queued up: to run the nspawn tests directly from the image file, without copying the files over. But I get filesystem corruption currently, so I'm leaving that part out for now.

poettering commented:
This looks pretty OK to me, but I can't claim I am a shell guru enough to be the ideal reviewer of this. But let's merge — as soon as the CI is happy?

keszybz commented:
I missed some new files during a rebase. Let's see how it goes now.

keszybz commented:
@mrc0mmand could you take a look? I filed https://github.com/systemd/systemd-centos-ci/pull/202, but I'm not sure if it does DTRT.

keszybz commented:
It turns out that the three tests failing on ubuntu autopkgtests had errors. They should be fixed now.

poettering commented:
hmm, needs rebase, too?

keszybz commented:
systemd.systemd failed with `[fuzz-unit-file-build]: meson.build:53:0: ERROR: Program or command '/home/appuser/fuzzer/tools/add-git-hook.sh' not found or not executable`. No idea, but doesn't look related.

centos ci (arch in kvm) failed with
```
11:43:30     archlinux_systemd_ci: FAILED: src/udev/v4l_id 
11:43:30     archlinux_systemd_ci: cc  -o src/udev/v4l_id 'src/udev/3eb56e7@@v4l_id@exe/v4l_id_v4l_id.c.o' -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-z,now -fstack-protector -Wl,--gc-sections -Wl,--start-group src/udev/libudev_static.a src/shared/libsystemd-shared-245.so src/udev/libudev-basic.a -Wl,--end-group '-Wl,-rpath,$ORIGIN/:$ORIGIN/../shared' -Wl,-rpath-link,/build/build/src/udev -Wl,-rpath-link,/build/build/src/shared
11:43:30     archlinux_systemd_ci: /usr/bin/ld: error: /build/build/src/udev/libudev.so.1: file too short
```
Looks like no disk space.

keszybz commented:
semaphoreci failed with a timeout during build too. Looks like the pandemic is affecting CI infra quality...

mrc0mmand commented:
> systemd.systemd failed with `[fuzz-unit-file-build]: meson.build:53:0: ERROR: Program or command '/home/appuser/fuzzer/tools/add-git-hook.sh' not found or not executable`. No idea, but doesn't look related.
> 
> centos ci (arch in kvm) failed with
> 
> ```
> 11:43:30     archlinux_systemd_ci: FAILED: src/udev/v4l_id 
> 11:43:30     archlinux_systemd_ci: cc  -o src/udev/v4l_id 'src/udev/3eb56e7@@v4l_id@exe/v4l_id_v4l_id.c.o' -Wl,--as-needed -Wl,--no-undefined -Wl,-z,relro -Wl,-z,now -fstack-protector -Wl,--gc-sections -Wl,--start-group src/udev/libudev_static.a src/shared/libsystemd-shared-245.so src/udev/libudev-basic.a -Wl,--end-group '-Wl,-rpath,$ORIGIN/:$ORIGIN/../shared' -Wl,-rpath-link,/build/build/src/udev -Wl,-rpath-link,/build/build/src/shared
> 11:43:30     archlinux_systemd_ci: /usr/bin/ld: error: /build/build/src/udev/libudev.so.1: file too short
> ```
> 
> Looks like no disk space.

Unfortunately, no, that's just NFS acting up, *sigh*... retriggered.

keszybz commented:
The tests were reworked to used a cached image. This generally make the whole thing more efficient, because we only build ~8 images, not 45 that were needed until now. But parallel running of tests does not work anymore. So I think the semaphore and centos tests are failing because they attempt to run the test in parallel.
@mrc0mmand what do you think? Should I try to restore the ability to run tests in parallel, or should the CI framework be reworked to run the tests serially? How much effective parallellism is there in our runners? If it is something like 2 or 4, then probably running things serially and building the images less times is going to be better. But if the we have a bazillion cores, then running things in parallel, even redundantly, might still be faster...

mrc0mmand commented:
> The tests were reworked to used a cached image. This generally make the whole thing more efficient, because we only build ~8 images, not 45 that were needed until now. But parallel running of tests does not work anymore. So I think the semaphore and centos tests are failing because they attempt to run the test in parallel.
> @mrc0mmand what do you think? Should I try to restore the ability to run tests in parallel, or should the CI framework be reworked to run the tests serially? How much effective parallellism is there in our runners? If it is something like 2 or 4, then probably running things serially and building the images less times is going to be better. But if the we have a bazillion cores, then running things in parallel, even redundantly, might still be faster...

In CentOS CI we have 8 cores, so I give each test 2 cores which results in 4-task parallelism. But honestly, the most wasteful part of the testsuite is the setup phase (building of the image), so in this case I'm all for ditching parallelism in favor of having to build less images (and I would be able to drop the [bash contraption](https://github.com/systemd/systemd-centos-ci/blob/master/common/task-control.sh#L153-L212) I wrote to handle all this stuff, which would make things more straightforward).

As for other CIs:
 * Sempahore apparently timed out during `meson test` (I think it doesn't run any integration tests). I suspect it's just a flake
 * Azure failed with `meson.build:53:0: ERROR: Program or command '/home/appuser/fuzzer/tools/add-git-hook.sh' not found or not executable` - I rescheduled it several times and it always fails with the same error, so it seems to be relevant (or at least it's consistent)
 * not sure what's going on with Ubuntu CI though (cc @ddstreet, @xnox)

keszybz commented:
See @mbiebl's comment https://github.com/systemd/systemd/pull/15170#issuecomment-602264477:
> the debian branches were renamed a while ago. Not sure why CI still references the old "experimental" branch. I quickly created a experimental branch to unbreak CI. Hopefully this helps.

keszybz commented:
> In CentOS CI we have 8 cores, so I give each test 2 cores which results in 4-task parallelism. But honestly, the most wasteful part of the testsuite is the setup phase (building of the image)

Right now 7 distinct images are built. So some parallelization would be effective. But maybe let's first get this running in non-parallel form. This should be simpler, and I'm not fully convinced yet that this new approach is what we want. Maybe we'll be able to enable better reuse of images, for example by running qemu in volatile snapshot mode.

How to handle the transition? Should I submit a PR for systemd-centos-ci to unparallelize?

mrc0mmand commented:
> Right now 7 distinct images are built. So some parallelization would be effective. But maybe let's first get this running in non-parallel form. This should be simpler, and I'm not fully convinced yet that this new approach is what we want. Maybe we'll be able to enable better reuse of images, for example by running qemu in volatile snapshot mode.
> 
> How to handle the transition? Should I submit a PR for systemd-centos-ci to unparallelize?

I'll prepare a PR which makes the parallelization temporarily conditioned (like don't parallelize if `grep IMAGE_NAME= test/test-functions` is true). This should help us to debug this properly here and not slow down current test runs (and it also won't break when this gets merged).

mrc0mmand commented:
Hm, not sure why, but on Arch Linux the `TEST-01-BASIC` fails because of:

```
    archlinux_systemd_ci: [0;38;5;245mActivating default unit: testsuite.target[0m
    archlinux_systemd_ci: [0;38;5;245mtestsuite.target: Failed to load configuration: No such file or directory[0m
    archlinux_systemd_ci: [0;1;31mUnit testsuite.target not found.[0m
    archlinux_systemd_ci: Falling back to rescue.target.
```

but on CentOS 7 it works as expected :thinking:

mrc0mmand commented:
Nevertheless, the CentOS 7 job finished successfully and seems to be only ~7 minutes slower than the parallelized job, which is not bad!

ddstreet commented:
> * not sure what's going on with Ubuntu CI though (cc @ddstreet, @xnox)

It seems like the Ubuntu CI configuration was changed somehow, e.g. using the older 'experimental' debian branch instead of master; and also it seems the wrong PPA was being used as well, instead of using "http://ppa.launchpad.net/upstream-systemd-ci/systemd-ci/ubuntu" some of the failing tests were using "http://ppa.launchpad.net/pitti/systemd-semaphore/ubuntu".

I think both of those have been corrected now, maybe @evverx or @mrc0mmand or @keszybz fixed up the Ubuntu CI config?

Unfortunately, it seems that wasn't the only issue, as since around Friday, the Ubuntu CI seem to be hanging/failing right after installing the newly-built systemd packages.  I can reproduce the failure on the production test instances, but not on a private test setup I have, so I'm working on getting access to the production test instances to try to debug.

ddstreet commented:
> > * not sure what's going on with Ubuntu CI though (cc @ddstreet, @xnox)
> 
> Unfortunately, it seems that wasn't the only issue, as since around Friday, the Ubuntu CI seem to be hanging/failing right after installing the newly-built systemd packages. I can reproduce the failure on the production test instances, but not on a private test setup I have, so I'm working on getting access to the production test instances to try to debug.

I'm cautiously optimistic that #15215 fixes the Ubuntu CI breakage...

keszybz commented:
I merged #15215 and restarted the tests here. Let's see.

evverx commented:
> I think both of those have been corrected now, maybe @evverx or @mrc0mmand or @keszybz fixed up the Ubuntu CI config?

The config was fixed a long time ago but it looks like the bash script that @keszybz seems to use to restart Ubuntu CI has never been updated. To judge from
![Screenshot 2020-03-24 at 20 11 03](https://user-images.githubusercontent.com/10195800/77456415-41279f00-6e0c-11ea-9159-208810e1c4e7.png)
it still points to the experimental branch and "pitti/systemd-semaphore" and doesn't pass CONFFLAGS_UPSTREAM=--werror -Dslow-tests=true.

evverx commented:
@mbiebl @ddstreet would it be possible to pass `--print-errorlogs` to `meson` in https://salsa.debian.org/systemd-team/systemd/-/blob/debian/master/debian/rules#L307 so that it would be easier to figure out why `test-fileio` failed there? Could you also somehow avoid running `cat build-deb/meson-logs/testlog.txt` when the tests are run upstream? Since it's too noisy I don't think it's particularly helpful. Plus due to some semaphore limitations, it can't capture all that stuff anyway so let's try to keep what's useful.

keszybz commented:
> The config was fixed a long time ago but it looks like the bash script that @keszybz seems to use to restart Ubuntu CI has never been updated.

Yep, my bad. I use a script I got from @martinpitt ages ago, and didn't think to update it. So most likely the whole mess with branch experimental is my fault.

> Could you also somehow avoid running `cat build-deb/meson-logs/testlog.txt` when the tests are run upstream?

+1

keszybz commented:
```
(Reading database ... 
(Reading database ... 5%
(Reading database ... 10%
(Reading database ... 15%
(Reading database ... 20%
(Reading database ... 25%
(Reading database ... 30%
(Reading database ... 35%
(Reading database ... 40%
(Reading database ... 45%
(Reading database ... 50%
(Reading database ... 55%
(Reading database ... 60%
(Reading database ... 65%
(Reading database ... 70%
(Reading database ... 75%
(Reading database ... 80%
(Reading database ... 85%
(Reading database ... 90%
(Reading database ... 95%
(Reading database ... 100%
(Reading database ... 60879 files and directories currently installed.)
```
What about that? It appears maybe a hundred times in each test log...

keszybz commented:
> why test-fileio failed

It failed because I'm an idiot. I chose `/proc/uptime` for the test which tries to write the file and falls back to checking if the content already matches. There's a 10 ms window in which it does, so it's interesting that the test passed in as many places as it did.

mrc0mmand commented on test/test-functions:
```diff
@@ -741,29 +915,24 @@ install_config_files() {
     > $initdir/etc/machine-id
+
+    # This is used by TEST-30-ONCLOCKCHANGE
+    inst_any /usr/share/zoneinfo/Europe/Kiev
```
Is this necessary, as we install the exact same files, among others, in `install_zoneinfo` below (which is also called by `setup_basic_environment`?

keszybz replied:
It was done in the original `TEST-30-*/test.sh`, so I assumed it serves a purpose ;) Gone now.

mrc0mmand commented on test/test-functions:
```diff
@@ -198,21 +300,28 @@ run_qemu() {
 console=$CONSOLE \
 selinux=0 \
 $_cgroup_args \
+SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-$1.units:/usr/lib/systemd/tests/testdata/units: \
```
For some strange reason this has no effect when running the integration tests on Arch Linux (or at least in the CentOS CI variant of it).

Logs from TEST-01-BASIC:

CentOS 7:
```
Kernel command line:  root=/dev/sda1 rw raid=noautodetect rd.luks=0 loglevel=2 init=/usr/lib/systemd/systemd console=ttyS0 selinux=0  SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-01.units:/usr/lib/systemd/tests/testdata/units: systemd.unit=testsuite.target systemd.wants=testsuite-01.service systemd.wants=end.service debug systemd.log_level=debug systemd.log_target=console 
...
Looking for unit files in (higher priority first):
        /usr/lib/systemd/tests/testdata/testsuite-01.units
        /usr/lib/systemd/tests/testdata/units
        /etc/systemd/system.control
        /run/systemd/system.control
        /run/systemd/transient
        /run/systemd/generator.early
        /etc/systemd/system
        /etc/systemd/system.attached
        /run/systemd/system
        /run/systemd/system.attached
        /run/systemd/generator
        /usr/local/lib/systemd/system
        /usr/lib/systemd/system
        /run/systemd/generator.late
```

Arch Linux (with sanitizers):
```
Kernel command line:  root=/dev/sda1 rw raid=noautodetect rd.luks=0 loglevel=2 init=/usr/lib/systemd/systemd-under-asan console=ttyS0 selinux=0  SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-01.units:/usr/lib/systemd/tests/testdata/units: systemd.unit=testsuite.target systemd.wants=testsuite-01.service systemd.wants=end.service  
...
Looking for unit files in (higher priority first):
        /etc/systemd/system.control
        /run/systemd/system.control
        /run/systemd/transient
        /run/systemd/generator.early
        /etc/systemd/system
        /etc/systemd/system.attached
        /run/systemd/system
        /run/systemd/system.attached
        /run/systemd/generator
        /usr/local/lib/systemd/system
        /usr/lib/systemd/system
        /run/systemd/generator.late
```

Arch Linux:
```
Kernel command line:  root=/dev/sda1 rw raid=noautodetect rd.luks=0 loglevel=2 init=/usr/lib/systemd/systemd console=ttyS0 selinux=0  SYSTEMD_UNIT_PATH=/usr/lib/systemd/tests/testdata/testsuite-01.units:/usr/lib/systemd/test
s/testdata/units: systemd.unit=testsuite.target systemd.wants=testsuite-01.service systemd.wants=end.service debug systemd.log_level=debug systemd.log_target=console 
...
Looking for unit files in (higher priority first):
        /etc/systemd/system.control
        /run/systemd/system.control
        /run/systemd/transient
        /run/systemd/generator.early
        /etc/systemd/system
        /etc/systemd/system.attached
        /run/systemd/system
        /run/systemd/system.attached
        /run/systemd/generator
        /usr/local/lib/systemd/system
        /usr/lib/systemd/system
        /run/systemd/generator.late
```

mrc0mmand replied:
Ah, I see. After some poking around, the issue is in the default initrd image.

By default, Arch Linux generates the initrd image with the `udev` hook (among others) included:

```
# grep ^HOOKS /etc/mkinitcpio.conf 
HOOKS=(base udev autodetect modconf block filesystems keyboard fsck)
```

However, in this case, `/sbin/init` is a symlink to a `busybox` binary, which then executes a simple init wrapper, that, in the end, executes systemd. Apparently, in this case, the set env variables from the kernel command line got lost around the way.

After switching from the `udev` hook to the `systemd` one (which [includes](https://wiki.archlinux.org/index.php/Mkinitcpio#Common_hooks) the `udev` one as well), everything works as expected and the test passes. I'll go ahead and update the CI scripts accordingly.

mrc0mmand replied:
Hopefully fixed by https://github.com/systemd/systemd-centos-ci/pull/230

mrc0mmand commented:
@keszybz one additional thing I noticed - at least in CentOS CI we run through all phases for each integration tests, like:

```bash
for test in test/TEST-??-*; do
    ...
    make -C $test clean setup run clean-again ...
done
```

I suspect we should drop the `clean` and `clean-again` phases so the image doesn't have to be constantly re-generated, right?

mrc0mmand commented:
Argh, this feature is dumb, sorry for the noise...

keszybz commented:
> @keszybz one additional thing I noticed - at least in CentOS CI we run through all phases for each integration tests, like:
> 
> ```shell
> for test in test/TEST-??-*; do
>     ...
>     make -C $test clean setup run clean-again ...
> done
> ```
> 
> I suspect we should drop the `clean` and `clean-again` phases so the image doesn't have to be constantly re-generated, right?

Yeah, I have some patches queued up to allow Ubuntu to just run our `run-integration-tests.sh` with some options, and they should work for that case too. But calling `clean` just makes things slow, it doesn't break stuff, so I thought I'd leave that for later.

mrc0mmand commented:
> Yeah, I have some patches queued up to allow Ubuntu to just run our `run-integration-tests.sh` with some options, and they should work for that case too. But calling `clean` just makes things slow, it doesn't break stuff, so I thought I'd leave that for later.

I see. As the runtime for the KVM job went from ~90 minutes to ~160 minutes, I went ahead and [tweaked](https://github.com/systemd/systemd-centos-ci/commit/fafd7e4e95e5a1a6c795cc746a8b2f7b9d2a9001) the no-parallelization workaround to skip the clean phase. Hopefully it's going to help a bit.

ddstreet commented:
> @mbiebl @ddstreet would it be possible to pass `--print-errorlogs` to `meson` in https://salsa.debian.org/systemd-team/systemd/-/blob/debian/master/debian/rules#L307 so that it would be easier to figure out why `test-fileio` failed there? Could you also somehow avoid running `cat build-deb/meson-logs/testlog.txt` when the tests are run upstream? Since it's too noisy I don't think it's particularly helpful. Plus due to some semaphore limitations, it can't capture all that stuff anyway so let's try to keep what's useful.

I've created a new debian branch 'upstream-ci':
https://salsa.debian.org/systemd-team/systemd/-/tree/upstream-ci

This is currently the same as 'debian/master', but intended to put upstream-specific changes into without affecting the master debian branch. 

@evverx can you (or another admin with test config access) change the debian branch that the Ubuntu CI tests use from 'debian/master' over to 'upstream-ci'?  That way we'll be able to easily add pretty much any needed upstream-specific test changes without delay.

I have this MR open to update the 'upstream-ci' branch with the change requested here:
https://salsa.debian.org/systemd-team/systemd/-/merge_requests/78
Once I have a chance to test it with an autopkgtest run, I'll push it into the 'upstream-ci' branch (assuming it tests ok).

@mbiebl this ^ is what we discussed in the libssl MR and earlier about an upstream-specific branch; let me know if you have any concerns.

ddstreet commented:
> ```
> (Reading database ... 
> (Reading database ... 5%
> (Reading database ... 10%
> (Reading database ... 15%
> (Reading database ... 20%
> (Reading database ... 25%
> (Reading database ... 30%
> (Reading database ... 35%
> (Reading database ... 40%
> (Reading database ... 45%
> (Reading database ... 50%
> (Reading database ... 55%
> (Reading database ... 60%
> (Reading database ... 65%
> (Reading database ... 70%
> (Reading database ... 75%
> (Reading database ... 80%
> (Reading database ... 85%
> (Reading database ... 90%
> (Reading database ... 95%
> (Reading database ... 100%
> (Reading database ... 60879 files and directories currently installed.)
> ```
> 
> What about that? It appears maybe a hundred times in each test log...

yeah, this (and really the rest of the test instance setup, i.e. apt update and installing dependencies, etc) adds tons of noise into the logs.  I do want to figure out some way to remove this or at least only print it on error, but not sure when exactly I'll be able to get to it - hopefully sometime soon.  I might be able to add a --quiet param in some places to at least make the log output shorter, though I don't think it would eliminate it completely, but that may be a quick improvement I'll look into asap.

keszybz commented:
I reworked the loop device handling a bit. losetup is always followed by udevadm settle now, and the device is cleaned up using `trap`. I hope this will fix the issues seen in ubuntuautopkg tests where the loopdevice partitions were missing.

mrc0mmand commented:
After dropping the cleanup phases the "overhead" from dropping parallelism is now 10 - 15 minutes. In the CentOS 7  job it's surprisingly not noticeable at all.

In the sanitizers job there's now a suspicious memory leak in TEST-47-REPART, but this needs a further investigation, as the stack trace is useless without a working symbolizer:

```
[   36.997470] testsuite-45.sh[361]: + sfdisk -d /tmp/zzz
[   37.042240] testsuite-45.sh[362]: + grep -v -e sector-size -e '^$'
[[0;32m  OK  [0m] Found device [0;1;39m/dev/ttyS0[0m.
[   38.703111] testsuite-45.sh[361]: =================================================================
[   38.714255] testsuite-45.sh[361]: ==361==ERROR: LeakSanitizer: detected memory leaks
[   38.723924] testsuite-45.sh[361]: Direct leak of 32 byte(s) in 1 object(s) allocated from:
[   38.728473] testsuite-45.sh[361]:     #0 0x7fb59ba72b11 in calloc (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xebb11)
[   38.734267] testsuite-45.sh[361]:     #1 0x7fb59b92f631 in fdisk_new_table (/usr/lib/libfdisk.so.1+0x19631)
[   38.742953] testsuite-45.sh[361]: SUMMARY: AddressSanitizer: 32 byte(s) leaked in 1 allocation(s).
[   38.871298] testsuite-45.sh[364]: ==364==WARNING: Symbolizer was blocked from starting itself!
[   38.927319] testsuite-45.sh[364]: =================================================================
[   38.931752] testsuite-45.sh[364]: ==364==ERROR: LeakSanitizer: detected memory leaks
[   38.935800] testsuite-45.sh[364]: Direct leak of 67584 byte(s) in 1 object(s) allocated from:
[   38.943281] testsuite-45.sh[364]:     #0 0x7fdc038b4959  (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xeb959)
[   38.949968] testsuite-45.sh[364]:     #1 0x7fdbff482e16  (/usr/bin/../lib/libLLVM-9.so+0xb06e16)
[   38.957617] testsuite-45.sh[364]:     #2 0x3fffffffe  (<unknown module>)
[   38.979253] testsuite-45.sh[364]:     #3 0x7ffff87b8e20  ([stack]+0x20e20)
[   39.002106] testsuite-45.sh[364]:     #4 0x79732d6d766c6c2e  (<unknown module>)
[   39.014749] testsuite-45.sh[364]: SUMMARY: AddressSanitizer: 67584 byte(s) leaked in 1 allocation(s).
[   40.535422] testsuite-45.sh[376]: LLVMSymbolizer: error reading file: No such file or directory
[   40.549165] testsuite-45.sh[351]: =================================================================
[   40.552768] testsuite-45.sh[351]: ==351==ERROR: LeakSanitizer: detected memory leaks
[   40.556016] testsuite-45.sh[351]: Direct leak of 32 byte(s) in 1 object(s) allocated from:
[   40.558899] testsuite-45.sh[351]:     #0 0x7f4116635959 in malloc (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xeb959)
[   40.562113] testsuite-45.sh[351]:     #1 0x5606d3e807e0 in xmalloc (/usr/bin/bash+0x817e0)
[   40.565634] testsuite-45.sh[351]: Indirect leak of 136 byte(s) in 4 object(s) allocated from:
[   40.569893] testsuite-45.sh[351]:     #0 0x7f4116635959 in malloc (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xeb959)
[   40.574519] testsuite-45.sh[351]:     #1 0x5606d3e807e0 in xmalloc (/usr/bin/bash+0x817e0)
[   40.578302] testsuite-45.sh[351]: Indirect leak of 56 byte(s) in 2 object(s) allocated from:
[   40.582304] testsuite-45.sh[351]:     #0 0x7f4116635959 in malloc (/usr/lib/clang/9.0.1/lib/linux/libclang_rt.asan-x86_64.so+0xeb959)
[   40.588368] testsuite-45.sh[351]:     #1 0x5606d3e807e0 in xmalloc (/usr/bin/bash+0x817e0)
[   40.592345] testsuite-45.sh[351]:     #2 0x7ffc14ba87bf  ([stack]+0x1e7bf)
[   40.595527] testsuite-45.sh[351]: SUMMARY: AddressSanitizer: 224 byte(s) leaked in 7 allocation(s).
[[0;1;31mFAILED[0m] Failed to start [0;1;39mTEST-45-REPARTn[0m.```

keszybz commented:
On Fri, Mar 27, 2020 at 08:32:17AM -0700, Dan Streetman wrote:
> @evverx can you (or another admin with test config access) change the debian branch that the Ubuntu CI tests use from 'debian/master' over to 'upstream-ci'?  That way we'll be able to easily add pretty much any needed upstream-specific test changes without delay.

Done. I hope I got it right. I changed all instances of %fmaster to %fupstream-ci.

Zbyszek

keszybz commented:
> In the sanitizers job there's now a suspicious memory leak in TEST-47-REPART, but this needs a further investigation, as the stack trace is useless without a working symbolizer:

Yeah, I don't really see how the changes in this PR would have influence over this.
But overall, I think we should just convert TEST-45 to a unit test. There's really no need to create an image to run a test that mangles a file in /tmp. I don't want to touch this branch now, because ubuntuautopkgtests are still running and I want to see if they pass.

keszybz commented:
> But overall, I think we should just convert TEST-45 to a unit test.

Definitely. It now takes 0.23 s. I'll submit this as part of the next PR.

All tests pass except for the Centos ci (arch in kvm with sanitizers) — the issue with TEST-45, and azure systemd.systemd — the issue with missing add-git-hook.sh. I think we can solve those later.
@fsumsal press the big green button please?

evverx commented:
> All tests pass except for the Centos ci (arch in kvm with sanitizers) — the issue with TEST-45

I think the issue that was addressed in 9a1862bfa6cdd1352101b232 has been reintroduced. With `testsuite.service` gone `LD_PRELOAD` is no longer reset where it should be reset.

> and azure systemd.systemd — the issue with missing add-git-hook.sh

According to `git bisect`, the commit where TEST-35-NETWORK-GENERATOR was converted into a unit test is the culprit.

> I think we can solve those later.
@fsumsal press the big green button please?

The point of CI among other things is to prevent PRs breaking it from being merged so I'm not sure it's a good idea to land this PR.

evverx commented:
> this ^ is what we discussed in the libssl MR and earlier about an upstream-specific branch; let me know if you have any concerns.

@ddstreet I'm wondering how you're planning to keep those branches in sync? Could it somehow be configurable on the debian package side without touching the webhooks via reading a config file that can be kept in the systemd repository anyone can change (by opening PRs triggering Ubuntu CI and semaphore CI to make sure they work)? The way this is currently dealt with is messy to be honest and generally explains why currently Ubuntu CI points to a non-existent branch and Semaphore CI is forgotten (and it doesn't seem to be the first time we can't seem to switch branches in one fell swoop).

keszybz commented:
> I think the issue that was addressed in [9a1862b](https://github.com/systemd/systemd/commit/9a1862bfa6cdd1352101b232a24f039b88ef4cee) has been reintroduced. With `testsuite.service` gone `LD_PRELOAD` is no longer reset where it should be reset.

Indeed. 9a1862bfa6cdd1352101b232a24f039b88ef4cee was using `testsuite.service.d`, but each test has a different name now. I adjusted the dropin to `testsuite-.service.d` now.

> According to git bisect, the commit where TEST-35-NETWORK-GENERATOR was converted into a unit test is the culprit.

I have no idea how this could be related.

> Done. I hope I got it right. I changed all instances of %fmaster to %fupstream-ci.

It seems I got it wrong. It left the "debian/" part in, and it seems it should be removed too.

evverx commented:
> I have no idea how this could be related.

I have to admit I have no idea either. All I know is that the issue goes away once that commit is reverted. Looks like a meson bug but I'm not sure.

keszybz commented:
```
+ pip3 install meson
Collecting meson
  Downloading https://files.pythonhosted.org/packages/49/d7/38a71de6adaa7a25a72587e69429ce78e137f4d545c7516341aa081f43e7/meson-0.53.2.tar.gz (1.6MB)
Building wheels for collected packages: meson
  Running setup.py bdist_wheel for meson: started
  Running setup.py bdist_wheel for meson: finished with status 'error'
  Complete output from command /usr/bin/python3 -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-cxk03cdj/meson/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" bdist_wheel -d /tmp/tmpzrbrlu1ppip-wheel- --python-tag cp36:
  usage: -c [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: -c --help [cmd1 cmd2 ...]
     or: -c --help-commands
     or: -c cmd --help
  
  error: invalid command 'bdist_wheel'
  
  ----------------------------------------
  Failed building wheel for meson
  Running setup.py clean for meson
Failed to build meson
Installing collected packages: meson
  Running setup.py install for meson: started
    Running setup.py install for meson: finished with status 'done'
Successfully installed meson-0.53.2
```
Hmmmmmm.

evverx commented:
This happens because `python3-wheel` isn't installed there. It's been failing this way for more than a year so I'm pretty sure it has nothing to do with `meson` failing to find `add-git-hook.sh` and certainly can be fixed later.

keszybz commented:
@evverx it seems you can run the tests locally (since you did a bisect). Can you post a quick howto? I tried running the test locally, but it immediately fails, so I think I'm doing something trivial wrong.

evverx commented:
To reproduce it locally I ran the following commands from `travis-ci/managers/fuzzbuzz.sh`:
```sh
wget https://app.fuzzbuzz.io/releases/cli/latest/linux/fuzzbuzz
chmod +x fuzzbuzz
./fuzzbuzz validate
./fuzzbuzz target test fuzz-unit-file --all
```
`fuzzbuzz` uses `docker` under the hood so it should be installed first:
https://docs.docker.com/install/linux/docker-ce/fedora/
https://docs.docker.com/install/linux/linux-postinstall/

`selinux` should be turned off as well to make it work on Fedora.

evverx commented:
It seems `docker` can't handle `test/testdata` (which is a symlink to `.`) introduced in that commit, which, in turn, leads to a build directory where `tools` is absent. Even though I decided to turn off FuzzBuzz in https://github.com/systemd/systemd/pull/15263 (which should make the CI green here) I'm not sure symlinks to directories should be kept in the repositories because they aren't always supported by various tools that could be used to build systemd.

keszybz commented:
fuzzbuzz was disabled in #15263. So maybe let's merge this now? @fsumsal, @poettering ?

mrc0mmand commented:
Given that Travis passed (results just didn't propagate here for whatever reason) and the Ubuntu CI fail looks unrelated (timeout) I'd say this PR is ready to go, thanks!

evverx commented:
@ddstreet @mbiebl Regarding "upstream-ci" and "debian/master", judging by https://github.com/systemd/systemd/pull/15333, the tests aren't run when "debian/master" is updated (which I think is more or less fine considering we no longer use it) but as much as I'm glad we often can catch downstream issues early, it would be great if you could somehow make sure "upstream-ci" isn't broken.
